dspace-api/src/main/java/org/dspace/discovery/indexobject/LDNMessageEntityIndexFactoryImpl.java

```diff
@@ -0,0 +1,156 @@
+import org.dspace.content.Item;
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Context;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Factory implementation implementation for the
+ * {@link IndexableLDNNotification}
+ *
+ * @author Stefano Maffei at 4science.com
+ */
+public class LDNMessageEntityIndexFactoryImpl extends IndexFactoryImpl<IndexableLDNNotification, LDNMessageEntity> {
+
+    @Autowired(required = true)
+    private LDNMessageService ldnMessageService;
+    @Autowired(required = true)
+    private ItemService itemService;
+
+    @Override
+    public Iterator<IndexableLDNNotification> findAll(Context context) throws SQLException {
+        final Iterator<LDNMessageEntity> ldnNotifications = ldnMessageService.findAll(context).iterator();
+        return new Iterator<>() {
+            @Override
+            public boolean hasNext() {
+                return ldnNotifications.hasNext();
+            }
+
+            @Override
+            public IndexableLDNNotification next() {
+                return new IndexableLDNNotification(ldnNotifications.next());
+            }
+        };
+    }
+
+    @Override
+    public String getType() {
+        return IndexableLDNNotification.TYPE;
+    }
+
+    @Override
+    public Optional<IndexableLDNNotification> findIndexableObject(Context context, String id) throws SQLException {
+        final LDNMessageEntity ldnMessage = ldnMessageService.find(context, id);
+        return ldnMessage == null ? Optional.empty() : Optional.of(new IndexableLDNNotification(ldnMessage));
+    }
+
+    @Override
+    public boolean supports(Object object) {
+        return object instanceof LDNMessageEntity;
+    }
+
+    @Override
+    public List<IndexableLDNNotification> getIndexableObjects(Context context, LDNMessageEntity object)
+        throws SQLException {
+        return Arrays.asList(new IndexableLDNNotification(object));
+    }
+
+    @Override
+    public SolrInputDocument buildDocument(Context context, IndexableLDNNotification indexableObject)
+        throws SQLException, IOException {
+        // Add the ID's, types and call the SolrServiceIndexPlugins
+        final SolrInputDocument doc = super.buildDocument(context, indexableObject);
+        final LDNMessageEntity ldnMessage = indexableObject.getIndexedObject();
+        // add schema, element, qualifier and full fieldName
+        doc.addField("notification_id", ldnMessage.getID());
+        doc.addField("queue_status_i", ldnMessage.getQueueStatus());
+        doc.addField("queue_status_s", LDNMessageEntity.getQueueStatus(ldnMessage));
+        addFacetIndex(doc, "queue_status", String.valueOf(ldnMessage.getQueueStatus()),
+            LDNMessageEntity.getQueueStatus(ldnMessage));
+        if (ldnMessage.getObject() == null || ldnMessage.getObject().getID() == null) {
+            throw new SQLException("ldnMessage " + ldnMessage.getID() + " is linked to no Item.");
```

Comment:
Does this need to throw an error?  Or should it just be a warning in the logs?  The reason I ask is that we are dependent on external services to send us **valid LDN messages**.  If they send us an invalid message without an "object", then that will cause an indexing error.

I'd much rather this be logged as a warning (`log.warn`), as we cannot have invalid LDN messages causing errors in indexing.  DSpace needs to not throw major errors if it's sent an invalid message... instead, we should just ignore that message.

Reply:
Comparing to all other FactoryIndexes on their _buildDocument(...)_ methods: everytime the code looks for a value but it is not found - the code just skips adding all the linked properties. We already do it for attributes like _origin_ and _target_ few lines below. It is correct to do the same on both the code portions you reported here.

Reply:
@frabacche : That's fine with me.  My main point is that we cannot throw **errors** in these index scripts.. that causes everything to fail.  Your new code looks better to me though I noticed a few small typos/bugs as I noted below in my recent code comments.